Use 'raw' assembly in main source tree

hanno-becker opened the pull request:
This PR modifies the way backends are developed and consumed.

Originally, we had AArch64 clean and optimized backends in `mlkem/native/aarch64`. Recently, we moved the clean backend to `test/aarch64_clean`, but kept the optimized backend.

Now some potential consumers, e.g. AWS-LC, need 'simplified' assembly without register macros or aliases. Previous PRs have introduced the `simpasm` script to provide this simplification. Yet, the simplified files are not part of the source tree, so when importing the library one has to call `simpasm`, which -- to cover the whole assembly tree -- needs to use both a native and a cross-compiler. This is inconvenient.

Moreover, we did not have a mechanism to install the `aarch64_clean` backend since its move to `test/aarch64_clean`, other than copy-paste -- not the end of the world, but not convenient either.

This commit moves both the clean and optimized aarch64 to a new directory `dev/`. What is shipped in the main source tree `mlkem/*`, in turn, it merely the _simplified_ assembly. The `autogen` script ensures that the simplified assembly files and the SLOTHY-optimized ones are in sync. 

What's more, `autogen` can now be called with `--aarch64-clean`, which similarly established `mlkem/native/aarch64/*` based on the _clean_ assembly -- this provides a convenient mechanism to test changes made to the clean backend. Also, `autogen` offers `--no-simplify`, which skips assembly simplification.

hanno-becker committed:
Move test/aarch64_clean -> dev/aarch64_clean

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
Add copies of assembly backends in dev/

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
Auto-generate simplified assembly in mlkem/

This commit replaces the shipped assembly in mlkem/* by its simplified
version not using register aliases or assembly macros.

The simplified assembly is auto-generated using `autogen` (and, under
the hood, `simpasm`) from the assembly in `dev/*`.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
autogen: Synchronize non-assembly files in dev/* with backends

Previous commits autogenerate the *.S files in the backends in mlkem/*,
while all other backend files are unsynchronized.

This commit modifies `autogen` to synchronize (by copy) all non *.S
backend files between `dev/*` and `mlkem/*`.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
Rename backend metadata headers to "meta.h"

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
autogen: Allow importing 'clean' AArch64 backend

With `autogen --aarch64-clean`, the 'clean' AArch64 backend will be
imported from `dev/aarch64_clean/*` into `mlkem/native/aarch64/*`.

With `autogen --no-simplify` -- which can be combined with `--aarch64-clean`,
simplification can be skipped.

Care has to be taken to not use `git ls-files` to list source files,
since that's only updated once files are checked in. That does not
work for `autogen`, which needs to create files and modify them afterwards
(e.g. when copying from `dev/*` to `mlkem/*` and then adjusting headers),
or when deleting files and then of course not modifying them anymore.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
autogen: Autogenerate header guards in dev/*.h

Previously, autogen would only autogenerate headers for files in `mlkem/*.h`.
This commit adjusts it to also generate header guards for `dev/*.h`. The files
are adjusted accordingly.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

mkannwischer commented on proofs/hol_light/arm/mlkem/mlkem_intt.S:
Why do we need to keep a copy of this file for the HOL Light Proof? Can't we use the simplified assembly directly? If that is the case I would prefer not duplicating the files.

hanno-becker replied:
The HOL Light ASM file is completely standalone and does not even have the header anymore. We could change that, but I would not want to do it in this PR.

mkannwischer replied:
Okay - would be nice to have only one file.

mkannwischer commented on README.md:
```diff
@@ -80,7 +80,7 @@ offers three backends for C, AArch64 and x86_64 - if you'd like contribute new b
 PR.
 
 Our AArch64 assembly is developed using [SLOTHY](https://github.com/slothy-optimizer/slothy): We write
-'clean' assembly by hand and automate micro-optimizations (e.g. see the [clean](test/aarch64_clean/src/ntt_clean.S)
+'clean' assembly by hand and automate micro-optimizations (e.g. see the [clean](dev/aarch64_clean/src/ntt_clean.S)
```
Please add a link to `dev/README.md` for those interest in how the final assembly was produced.

hanno-becker replied:
Ok

hanno-becker replied:
Done

mkannwischer commented on scripts/autogen:
Not related to your PR (sorry!): 
Currently linting errors with 
```
Autogenerated file proofs/hol_light/arm/mlkem/mlkem_ntt.S needs updating. Have you called scripts/autogenerated.py?
Writing new version to proofs/hol_light/arm/mlkem/mlkem_ntt.S.new
```

1) Could you adjust that error message `scripts/autogenerated.py` is now `scripts/autogen` (two places where this needs to change). 

2) Why is linting writing `proofs/hol_light/arm/mlkem/mlkem_ntt.S.new`? Linting should just error in case files are outdated.

hanno-becker replied:
1. Ok
2. To ease debugging

mkannwischer commented on scripts/autogen:
```diff
@@ -772,20 +771,66 @@ def check_asm_register_aliases():
+    if native_arch != source_arch:
+        if platform.system() == "Darwin":
+            print(
+                f"WARNING: Skipping simplification of file {infile} -- "
```
Discussions with @potsrevennil yesterday suggest that this limitation can be easily overcome. Please open an issue referencing this line so we do not forget to address this.

hanno-becker replied:
Changed the behaviour here to print the warning whenever a cross-chain is unavailable, not specifically when it's Darwin.

mkannwischer reviewed (changes requested):
Thanks @hanno-becker!

1) Linting fails with 
```
  Autogenerated file mlkem/native/aarch64/src/arith_native_aarch64.h needs updating. Have you called scripts/autogenerated.py?
  Writing new version to mlkem/native/aarch64/src/arith_native_aarch64.h.new
  5,6c5,6
  < #ifndef MLKEM_NATIVE_NATIVE_AARCH64_SRC_ARITH_NATIVE_AARCH64_H
  < #define MLKEM_NATIVE_NATIVE_AARCH64_SRC_ARITH_NATIVE_AARCH64_H
  ---
  > #ifndef MLKEM_NATIVE_DEV_AARCH64_OPT_SRC_ARITH_NATIVE_AARCH64_H
  > #define MLKEM_NATIVE_DEV_AARCH64_OPT_SRC_ARITH_NATIVE_AARCH64_H
  77c77
  < #endif /* MLKEM_NATIVE_NATIVE_AARCH64_SRC_ARITH_NATIVE_AARCH64_H */
  ---
  > #endif /* MLKEM_NATIVE_DEV_AARCH64_OPT_SRC_ARITH_NATIVE_AARCH64_H */
```
Running `autogen` locally (on MacOS) does not resolve this problem - I don't understand why. 

But running autogen locally updates mlkem/fips202/native/x86_64/meta.h - which is not flagged by CI linting. Here is the patch: 
```
diff --git a/mlkem/fips202/native/x86_64/meta.h b/mlkem/fips202/native/x86_64/meta.h
index 88f42739..f077bc75 100644
--- a/mlkem/fips202/native/x86_64/meta.h
+++ b/mlkem/fips202/native/x86_64/meta.h
@@ -3,8 +3,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef MLKEM_NATIVE_FIPS202_NATIVE_X86_64_XKCP_H
-#define MLKEM_NATIVE_FIPS202_NATIVE_X86_64_XKCP_H
+#ifndef MLKEM_NATIVE_FIPS202_NATIVE_X86_64_META_H
+#define MLKEM_NATIVE_FIPS202_NATIVE_X86_64_META_H
 /* Default FIPS202 assembly profile for AArch64 systems */
 
 #ifdef MLKEM_NATIVE_FIPS202_PROFILE_H
@@ -26,4 +26,4 @@
 
 #endif /* MLKEM_NATIVE_FIPS202_PROFILE_H */
 
-#endif /* MLKEM_NATIVE_FIPS202_NATIVE_X86_64_XKCP_H */
+#endif /* MLKEM_NATIVE_FIPS202_NATIVE_X86_64_META_H */
```

hanno-becker committed:
CI: Add dedicated job for running autogen

The `lint` job includes `autogen`, which includes the simplification
of assembly. Since we have assembly files for multiple architectures,
however, full assembly simplification requires a cross toolchain.

To avoid requiring the full-blown `ci-cross` shell for the `lint` job,
this commit changes `autogen` so that it skips over cross-simplification
unless the `--force-cross` parameter is given.
The `lint` job is then called without cross simplification, but instead
run on both an aarch64 and x86_64 machine, which jointly cover the entire
assembly.

As an additional safeguard, and also to check that simplification via
cross-toolchains work, a separate CI job for the checking of autogenerated
files is introduced which does run in the ci-cross shell and simplifies
all assembly.

The previous `simpasm` CI job is no longer needed (since the simplified
asm is now part of the source tree) and removed. Instead, a job is added
which reinstates the clean/opt backend with/without simplification, and
tests it.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
autogen: Adjust header guards at time of update-via-copy

Previously, `autogen` would update the backend headers in `mlkem/*`
in a two-step fashion: First, synchronize them with the headers in `dev/*`,
and then update the guards. This approach works in a plain `autogen`, since
_overall_, the file in `mlkem/*` should not change if it was already up to
date. However, in a `autogen --dry-run` run, it will fail in the interim
stage where the header from `dev/*` is copied but the header guard is not
yet adjusted.

This commit fixes this by adjusting the header guard as part of the
synchronization from `dev/*` to `mlkem/*`.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

hanno-becker committed:
README: Add reference to dev/README.md

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

mkannwischer commented on scripts/autogen:
```diff
@@ -794,22 +848,130 @@ def gen_hol_light_asm_file(infile, outfile, dry_run=False):
+def update_via_remove(filename, dry_run=False):
+    if dry_run is True:
+        print(
+            f"Autogenerated file {filename} needs removing. Have you called scripts/autogenerated.py?",
```
change to autogen, please

hanno-becker replied:
done

mkannwischer reviewed (approved):
Thanks @hanno-becker - I tested that the autogen script works with the different flag variations (`autogen --aarch64-clean`, `autogen --aarch64-clean --no-simplify`). 

One nit and then I'm happy with this PR.
